### PR TITLE
Detect dev session takeover

### PR DESCRIPTION
**Pull request by dmerand**
## Summary

Implements client-side detection and handling for multi-user dev session conflicts, building on backend changes in https://github.com/shop/world/pull/383392.

When multiple developers run `shopify app dev` on the same app/shop combination, the CLI now provides clear feedback to both users about the session takeover and gracefully handles the conflict.

### GraphQL Query Updates
- **dev-session-create.graphql**: Now requests `devSession`, `warnings`, and `userErrors` fields
- **dev-session-update.graphql**: Now requests `devSession` with user and websocket information

### Schema Updates
Updated local schema file (`app_dev_schema.graphql`) to match backend changes:
- Added `devSession` field to `DevSessionCreatePayload`
- Added `warnings` array to `DevSessionCreatePayload`
- Added `devSession` field to `DevSessionUpdatePayload`
- Added new types: `DevSessionWarning` and `DevSessionWarningCode` enum

### Implementation (dev-session.ts)

**Session State Tracking:**
- Added `DevSessionState` interface to track websocket URL and user information
- Added `currentSessionState` instance variable to maintain session metadata

**Create-Time Warning Display:**
- Stores initial session state when dev session is created
- Displays warnings from backend (e.g., `SESSION_TAKEOVER`)
- Warnings are non-blocking - session creation succeeds even with warnings

**Update-Time Takeover Detection:**
- Compares returned `websocketUrl` against expected local websocket URL on every update
- If mismatch detected → another developer has taken over the session
- Displays clear error message including the new user's email
- Throws `AbortError` to gracefully terminate the dev session with actionable next steps

### Scenario: User B Takes Over User A's Session

**User A starts dev:**
```
$ shopify app dev
✅ Ready, watching for changes in your app
```

**User B starts dev (same app/shop):**
```
$ shopify app dev
⚠️  Another user is already running a dev preview for this app on example.myshopify.com.
✅ Ready, watching for changes in your app
```

**User A makes a code change:**
```
[app-extension] Extension updated

❌ Error
└  
⚠️  Another developer (user-b@shopify.com) has taken over this dev session.
Your preview is no longer active. Terminating dev session...

Dev session has been taken over by another user.
You can restart by running \`shopify app dev\` again.
```

## Benefits

- ✅ **Immediate feedback for aggressor**: User B knows they displaced someone
- ✅ **Quick discovery for victim**: User A learns about takeover on next code change (typically seconds)
- ✅ **Clean separation**: Warnings (non-blocking) vs errors (blocking)
- ✅ **Graceful termination**: Clear messaging with actionable next steps
- ✅ **No false positives**: Only checks during actual mutations, not on normal disconnects
- ✅ **Future-proof**: Uses structured warning codes for extensibility

## Dependencies

This PR depends on backend changes:
- https://github.com/shop/world/pull/383392 - Adds session takeover warning on create

## Testing Plan

Once the backend PR is merged:
1. Test session takeover warning appears when User B starts dev
2. Test User A's session terminates gracefully on next update after takeover
3. Verify error messages are clear and actionable
4. Test edge cases:
   - Same user, multiple machines
   - Same user, same machine (should show no warning)
   - Rapid takeover scenarios

Addresses https://github.com/shop/issues-develop/issues/21606

### Measuring impact

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes

### CODE REVIEW

**Comment by dmerand** [comment]
* **#6799** <a href="https://app.graphite.com/github/pr/Shopify/cli/6799?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/Shopify/cli/6799?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts [hunk 1]
```diff
@@ -402,6 +409,35 @@ export class DevSession {
   private async devSessionUpdateWithRetry(payload: DevSessionUpdateOptions): Promise<DevSessionResult> {
     const result = await this.options.developerPlatformClient.devSessionUpdate(payload)
     const errors = result.devSessionUpdate?.userErrors ?? []
+    const devSession = result.devSessionUpdate?.devSession
+
+    // Check for session takeover
+    if (devSession && this.currentSessionState) {
+      const expectedWebsocketUrl = getWebSocketUrl(this.options.url)
+      
+      if (devSession.websocketUrl !== expectedWebsocketUrl) {
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts]: Should we better check if the URL OR the user changed? If you are working with localhost, another user may have updated your session, but the URL could be the same.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts [hunk 2]
```diff
@@ -402,6 +409,35 @@ export class DevSession {
...
+        // Another user has taken over the session
+        const newUserEmail = devSession.user?.email ?? 'another user'
+        await this.logger.error(
+          `\n⚠️  Another developer (${newUserEmail}) has taken over this dev session.\n` +
+          `Your preview is no longer active. Terminating dev session...`
+        )
+        
+        // Throw AbortError to terminate gracefully
+        throw new AbortError(
+          'Dev session has been taken over by another user.',
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts]: It could actually be yourself from a different tab or device. What about a generic copy like `Your dev preview has been updated from another session`? Or terminal?

Also, externally we refer to this as `dev preview`.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts (hunk 1 above)
Comment (craigmichaelmartin) [reply packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts]: Good idea, updated!

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts (hunk 2 above)
Comment (craigmichaelmartin) [reply packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts]: Good point! Updated with new desired copy.

File: .eslintrc.cjs [hunk 3]
```diff
@@ -168,5 +168,5 @@ module.exports = {
       },
     },
   ],
-  ignorePatterns: [],
+  ignorePatterns: ['**/generated/'],
```
Comment (craigmichaelmartin) [inline .eslintrc.cjs]: I opted to ignore generated directories, rather than lint autogenerated type files.

Comment (craigmichaelmartin) [reply .eslintrc.cjs]: Just kidding. Apparently they need to be.

**Comment by craigmichaelmartin** [comment]
I left a [comment](https://shopify.slack.com/archives/C073MME9MFA/p1770412628744629) in slack that I'm surprised were doing this validation logic regarding dev preview takeover for updates in the CLI (vs the DevApi)? I expected the latter (server validation vs client validation reasons; + it would make tuning this logic/copy easier in the future since CLI is versioned) but stuck with how it was for now.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts [hunk 4]
```diff
@@ -401,6 +408,36 @@ export class DevSession {
...
+    if (devSession && this.currentSessionState) {
+      const expectedWebsocketUrl = getWebSocketUrl(this.options.url)
+      const expectedUserId = this.currentSessionState?.userId
+
+      if (devSession.websocketUrl !== expectedWebsocketUrl || devSession.user?.id !== expectedUserId) {
+        // Another user has taken over the session
+        const newUserEmail = devSession.user?.email
+        await this.logger.error(
+          `Another developer ${newUserEmail ? `(${newUserEmail}) ` : ''}has taken ownership of this dev preview.`,
+        )
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:423]: Why showing the error twice? I'd get rid of this if we are going to abort.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts [hunk 5]
```diff
@@ -401,6 +408,36 @@ export class DevSession {
...
+
+        // Throw AbortError to terminate gracefully
+        const message = `Another user${
+          newUserEmail ? ` (${newUserEmail})` : ''
+        } has taken ownership of this dev preview. Your preview is no longer active.`
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:428]: Not a big deal, but the issue I mentioned [here](https://github.com/Shopify/cli/pull/6799#discussion_r2754728371) is not fixed.

If the only change is the websocket URL, we are saying `Another user`, but showing the same email.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts [hunk 6]
```diff
@@ -414,7 +451,31 @@ export class DevSession {
...
+        userId: devSession.user?.id ?? null,
+        userEmail: devSession.user?.email ?? null,
+      }
+    }
+
+    // Display warnings (non-blocking)
+    if (warnings.length > 0) {
+      await Promise.all(
+        warnings.map((warning) => {
+          const message = warning.code === 'SESSION_TAKEOVER' ? `⚠️  ${warning.message}` : warning.message
```
Comment (gonzaloriestra) [inline packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:470]: Why not just adding the emoji to all the warnings?

```suggestion
          const message = `⚠️ ${warning.message}`
```

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts (hunk 4 above)
Comment (craigmichaelmartin) [reply packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:423]: I think it's nice to see something in the log, even if we end up saying the same thing in the final error message. I asked about copy for these messages [here](https://shopify.slack.com/archives/C073MME9MFA/p1770411989219049) and while I didn't explicitly ask if both were warranted, there wasn't push back against having both.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts (hunk 5 above)
Comment (craigmichaelmartin) [reply packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:428]: Yep, you're right! Good call. That said, we've decided to just keep the same copy.

File: packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts (hunk 6 above)
Comment (craigmichaelmartin) [reply packages/app/src/cli/services/dev/processes/dev-session/dev-session.ts:470]: I guess because this was all that was specified to have it. We can revisit as we incorporate more warnings.
